DataDog/ddprof
# Free inconsistent dwfl modules

# What does this PR do?

- Free inconsistent Dwfl modules
We can come into situations where DSOs don't match the Dwfl objects.
In this situation we can free all of the PID objects to ensure we will reload all DSOs and all dwfl objects at the next sample.
- Add a unit test around the creation of dwfl modules
- Whole host mode bug : avoid loading the wrong file
When you have a local libc and a libc in the container, we were giving precedence to the local version of the file. This was a bad strategy that ended up loading inconsistent DSO <--> DWFL Modules

# Motivation

Looking at whole host profiling, we quickly encounter situations where we are not in sync between libdwfl objects and internal DSO objects.

# Additional Notes

There are still some limitations
- We should not compare dwfl and DSO objects using filenames. Perhaps storing file IDs would be more robust. I wanted to avoid storing pointers to the dwfl mod objects internally. Other ideas are welcome.
- When we encounter an error loading a mod, we attribute that to the file. This is not true in some cases. It should be attributed to the PID + file (and we could retry for a different PID). 

# How to test the change?

I added a unit test.
The best way to observe a change is to try out the whole host profiler in an environment profiling different languages.
